### PR TITLE
Product

### PR DESCRIPTION
**Pull request by groyvstreet**
Part of online-store microservice for Product entity (CRUD) created

### CODE REVIEW

File: ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/AddProductDto.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿namespace ModsenOnlineStore.Store.Domain.DTOs.ProductDTOs
+{
+    public class AddProductDto
+    {
+        public string Name { get; set; } = string.Empty;
+        public string Description { get; set; } = string.Empty;
+        public decimal Price { get; set; }
+        public decimal Discount { get; set; }
+        public int Quantity { get; set; }
+        public int ProductTypeId { get; set; }
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/AddProductDto.cs:15]: please add empty lines between properties, methods and so on in all classes

File: ModsenOnlineStore.Store.Domain/Entities/Product.cs [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
 namespace ModsenOnlineStore.Store.Domain.Entities
 {
     public class Product
     {
         public int Id { get; set; }
+        public string Name { get; set; } = string.Empty;
+        public string Description { get; set; } = string.Empty;
+        [Column(TypeName = "decimal(18,4)")]
+        public decimal Price { get; set; }
+        [Column(TypeName = "decimal(18,4)")]
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Domain/Entities/Product.cs]: use fluent api and entity type configurations in such cases

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        public async Task<Product?> GetProductById(int id)
+        {
+            return await context.Products.AsNoTracking().FirstOrDefaultAsync(p => p.Id == id);
+        }
+
+        public async Task<List<Product>> AddProduct(Product product)
+        {
+            context.Products.Add(product);
+            await context.SaveChangesAsync();
+            return await GetAllProducts();
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: please use empty line before 'return' statements

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        }
+
+        public async Task<List<Product>> UpdateProduct(Product product)
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: why do you return all products after update

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            context.Products.Update(product);
+            await context.SaveChangesAsync();
+            return await GetAllProducts();
+        }
+
+        public async Task<List<Product>> RemoveProductById(int id)
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: the same

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        {
+            return await context.Products.AsNoTracking().ToListAsync();
+        }
+
+        public async Task<Product?> GetProductById(int id)
+        {
+            return await context.Products.AsNoTracking().FirstOrDefaultAsync(p => p.Id == id);
+        }
+
+        public async Task<List<Product>> AddProduct(Product product)
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: the same

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs [hunk 7]
```diff
@@ -0,0 +1,60 @@
+﻿using AutoMapper;
+using ModsenOnlineStore.Common;
+using ModsenOnlineStore.Store.Application.Interfaces.ProductInterfaces;
+using ModsenOnlineStore.Store.Domain.DTOs.ProductDTOs;
+using ModsenOnlineStore.Store.Domain.Entities;
+
+namespace ModsenOnlineStore.Store.Infrastructure.Services.ProductService
+{
+    public class ProductService : IProductService
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: such services should be placed in application layer, infrastructure layer should store logic of interacting with  external objects like db, file system and so on

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            return new ResponseInfo<List<GetProductDto>>(productDtos, true, "all products");
+        }
+
+        public async Task<ResponseInfo<GetProductDto>> GetProductById(int id)
+        {
+            var product = await repository.GetProductById(id);
+
+            if (product is null)
+            {
+                return new ResponseInfo<GetProductDto>(null, true, "product");
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: what is null, true, 'product' here? please add properties initialization or named parameters, so I do not need to remember constructor

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+            }
+
+            var productDto = mapper.Map<GetProductDto>(product);
+            return new ResponseInfo<GetProductDto>(productDto, true, "product");
+        }
+
+        public async Task<ResponseInfo<List<GetProductDto>>> AddProduct(AddProductDto putProductDto)
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: why do you return all products? same in update and delete

File: ModsenOnlineStore.Store.API/AutoMapperProfile.cs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+﻿using AutoMapper;
+using ModsenOnlineStore.Store.Domain.DTOs.ProductDTOs;
+using ModsenOnlineStore.Store.Domain.Entities;
+
+namespace ModsenOnlineStore.Store.API
+{
+    public class AutoMapperProfile : Profile
+    {
+        public AutoMapperProfile()
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.API/AutoMapperProfile.cs]: Name it ProductProfile as it should store only product related mapping configurations

File: ModsenOnlineStore.Store.API/Controllers/ProductController.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
+﻿using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using ModsenOnlineStore.Store.Application.Interfaces.ProductInterfaces;
+using ModsenOnlineStore.Store.Domain.DTOs.ProductDTOs;
+
+namespace ModsenOnlineStore.Store.API.Controllers
+{
+    [Route("[controller]")]
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.API/Controllers/ProductController.cs]: api/[controller]

File: ModsenOnlineStore.Store.API/Controllers/ProductController.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    [ApiController]
+    public class ProductController : ControllerBase
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.API/Controllers/ProductController.cs]: name controllers in plural form

File: ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/AddProductDto.cs (hunk 1 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/AddProductDto.cs:15]: fixed

File: ModsenOnlineStore.Store.Domain/Entities/Product.cs (hunk 2 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Domain/Entities/Product.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs (hunk 3 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs (hunk 4 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs (hunk 5 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs (hunk 6 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Data/ProductRepository.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs (hunk 7 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs (hunk 8 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: fixed

File: ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs (hunk 9 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Infrastructure/Services/ProductService/ProductService.cs]: fixed

File: ModsenOnlineStore.Store.API/AutoMapperProfile.cs (hunk 10 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.API/AutoMapperProfile.cs]: fixed

File: ModsenOnlineStore.Store.API/Controllers/ProductController.cs (hunk 11 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.API/Controllers/ProductController.cs]: fixed

File: ModsenOnlineStore.Store.API/Controllers/ProductController.cs (hunk 12 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.API/Controllers/ProductController.cs]: fixed

File: ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/GetProductDto.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+        public int Id { get; set; }
+
+        public string Name { get; set; } = string.Empty;
+
+        public string Description { get; set; } = string.Empty;
+
+        [Column(TypeName = "decimal(18,4)")]
+        public decimal Price { get; set; }
+
+        [Column(TypeName = "decimal(18,4)")]
```
Comment (elv1sabene) [inline ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/GetProductDto.cs]: do you need this attribute in this model? and another above

**Review by elv1sabene — CHANGES_REQUESTED** [review]

File: ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/GetProductDto.cs (hunk 13 above)
Comment (groyvstreet) [reply ModsenOnlineStore.Store.Domain/DTOs/ProductDTOs/GetProductDto.cs]: no i don't need. overlooked this moment

**Pull request merged**